app/lib/bulk_download_service.rb

```diff
@@ -57,22 +57,30 @@ def self.generate_curl_configuration(study_files:,
       hca_client = ApplicationController.hca_azul_client
       default_catalog = hca_client.default_catalog
       curl_configs << "-H \"Authorization: Bearer #{DataRepoClient.new.access_token['access_token']}\""
-      tdr_file_configs = tdr_files.map do |shortname, file_infos|
-        file_infos.map do |file_info|
-          if file_info['file_type'] == 'Project Manifest'
-            # generate manifest link using HCA project UUID from :url property
-            manifest = hca_client.get_project_manifest_link(default_catalog, file_info['url'])
-            file_config = "--location\n" # add location directive to allow following 302 redirect to manifest location
-            file_config += "url=\"#{manifest['Location']}\"\n"
-          else
-            file_config = "url=\"#{file_info['url']}\"\n"
+      tdr_file_configs = []
+      tdr_files.map do |shortname, file_infos|
+        # pull out project manifest and process separately
+        manifests, files = file_infos.partition { |f| f['file_type'] == 'Project Manifest' }
+        manifest_info = manifests.first
+        manifest = hca_client.get_project_manifest_link(default_catalog, manifest_info['url'])
+        manifest_config = "--location\n" # add location directive to allow following 302 redirect to manifest location
+        manifest_config += "url=\"#{manifest['Location']}\"\n"
+        manifest_config += "output=\"#{shortname}/#{manifest_info['name']}\""
+        tdr_file_configs << manifest_config
+        # now process remainder of analysis/sequence files for download in parallel to speed up process
+        Parallel.map(files, in_threads: 100) do |file_info|
+          begin
+            drs_info = ApplicationController.data_repo_client.get_drs_file_info(file_info[:drs_id])
```

Comment:
Please open a ticket upstream so we don't need to send a bunch of parallel requests.  Client multithreading is an API code smell -- if clients (like SCP) need to do it, it's a strong hint that the API needs improvement.

Naively, I suspect SCP could avoid parallelizing requests if Data Repo API had a version of the endpoint underlying `get_drs_file_info` that accepted a comma-delimited list of DRS IDs.  So the upstream ticket would be something like "Enable passing comma-delimited list of DRS IDs to \<API endpoint name\>".

That would allow client code like SCP's to be faster, simpler, and more robust.

Reply:
I actually did that months ago: https://broadworkbench.atlassian.net/browse/DR-1947.  Given the lack of movement on it, I doubt this is something they feel inclined to do unless we make a lot of noise, and I'm not sure this is the place we want to spend our capital getting things implemented.

Also, we had this exact same behavior for TDR files beforehand as well, just in a different place.  And we are doing the same thing for generating signed URLs for files in SCP study buckets on line 30 of this same service (though in that case, we're only generating signed URLs, and I don't think that emits an API request to GCS).  Is your objection to doing so here because it requires API calls?  And not just parallelizing method calls to an underlying SDK?

Reply:
Ah, thanks for noting the ticket.  I that could be noted in a code comment here, I promise not to suggest this next time!

> Is your objection to doing so here because it requires API calls? And not just parallelizing method calls to an underlying SDK?

My objection is parallelizing many network requests that could -- with a likely small amount of upstream effort -- be more quickly, reliably, and simply batched in a single request by all clients.  So if the API or SDK calls needs network I/O, and gets multi-threaded, then I ask myself "Why can't this be a single request?"  

Often, it's just a use case that hasn't been noted upstream.  I appreciate that that part's done, and agree that there are other improvements that would be even more helpful!